README.md
![](http://libsisimai.org/static/images/logo/sisimai-x01.png)

[![License](https://img.shields.io/badge/license-BSD%202--Clause-orange.svg)](https://github.com/sisimai/rb-Sisimai/blob/master/LICENSE)
[![Coverage Status](https://img.shields.io/coveralls/sisimai/rb-Sisimai.svg)](https://coveralls.io/r/sisimai/rb-Sisimai)
[![Build Status](https://travis-ci.org/sisimai/rb-Sisimai.svg?branch=master)](https://travis-ci.org/sisimai/rb-Sisimai) 
[![Codacy Badge](https://api.codacy.com/project/badge/grade/38340177e6284a65be69c0c7c3dc2b58)](https://www.codacy.com/app/azumakuniyuki/rb-Sisimai)
[![Ruby](https://img.shields.io/badge/ruby-v2.1.0--v2.4.0-red.svg)](https://www.ruby-lang.org/)
[![Gem Version](https://badge.fury.io/rb/sisimai.svg)](https://badge.fury.io/rb/sisimai)

- [**README-JA(日本語)**](README-JA.md)
- [What is Sisimai](#what-is-sisimai)
    - [Key features](#key-features)
    - [Command line demo](#command-line-demo)
- [Setting Up Sisimai](#setting-up-sisimai)
    - [System requirements](#system-requirements)
    - [Install](#install)
        - [From RubyGems](#from-rubygems)
        - [From GitHub](#from-github)
- [Usage](#usage)
    - [Basic usage](#basic-usage)
    - [Convert to JSON](#convert-to-json)
    - [Read bounce object](#read-bounce-object)
    - [Callback feature](#callback-feature)
    - [One-Liner](#one-liner)
    - [Output example](#output-example)
- [Sisimai Specification](#sisimai-specification)
    - [Differences between Ruby version and Perl version](#differences-between-ruby-version-and-perl-version)
    - [Other specification of Sisimai](#other-specification-of-sisimai)
- [Contributing](#contributing)
    - [Bug report](#bug-report)
    - [Emails could not be parsed](#emails-could-not-be-parsed)
- [Other Information](#other-information)
    - [Related sites](#related-sites)
    - [See also](#see-also)
- [Author](#author)
- [Copyright](#copyright)
- [License](#license)

What is Sisimai
===============================================================================
Sisimai is a Ruby library for analyzing RFC5322 bounce emails and generating
structured data from parsed results. The Ruby version of Sisimai is ported from
the Perl version of Sisimai at [github.com/sisimai/p5-Sisimai](https://github.com/sisimai/p5-Sisimai/).

![](http://libsisimai.org/static/images/figure/sisimai-overview-1.png)

Key Features
-------------------------------------------------------------------------------
* __Convert Bounce Mails to Structured Data__
  * Supported formats are Ruby(Hash, Array) and JSON(String)
* __Easy to Install, Use.__
  * gem install
  * git clone & make
* __High Precision of Analysis__
  * 2 times higher than bounceHammer
  * Support 22 known MTAs and 5 unknown MTAs
  * Support 22 major MSPs(Mail Service Providers)
  * Support 2 major Cloud Email Delivery Services(JSON format)
  * Support Feedback Loop Message(ARF)
  * Can detect 27 error reasons

Command line demo
-------------------------------------------------------------------------------
The following screen shows a demonstration of Sisimai at the command line using
Ruby(rb-Sisimai) and Perl(p5-Sisimai) version of Sisimai.
![](http://libsisimai.org/static/images/demo/sisimai-dump-01.gif)

Setting Up Sisimai
===============================================================================

System requirements
-------------------------------------------------------------------------------
More details about system requirements are available at
[Sisimai | Getting Started](http://libsisimai.org/en/start) page.


* [Ruby 2.1.0 or later](http://www.ruby-lang.org/)
  * [__Oj | The fastest JSON parser and object serializer__](https://rubygems.org/gems/oj)
* Also works on [JRuby 9.0.4.0 or later](http://jruby.org)
  * [__JrJackson | A mostly native JRuby wrapper for the java jackson json processor jar__](https://rubygems.org/gems/jrjackson)

Install
-------------------------------------------------------------------------------
### From RubyGems

```shell
$ sudo gem install sisimai
Fetching: sisimai-4.21.0.gem (100%)
Successfully installed sisimai-4.21.0
Parsing documentation for sisimai-4.21.0
Installing ri documentation for sisimai-4.21.0
Done installing documentation for sisimai after 6 seconds
1 gem installed
```

### From GitHub

```shell
$ cd /usr/local/src
$ git clone https://github.com/sisimai/rb-Sisimai.git
$ cd ./rb-Sisimai
$ sudo make depend install-from-local
gem install bundle rake rspec coveralls
...
4 gems installed
bundle exec rake install
sisimai 4.21.0 built to pkg/sisimai-4.21.0.gem.
sisimai (4.21.0) installed.
```

Usage
===============================================================================

Basic usage
-------------------------------------------------------------------------------
`make()` method provides feature for getting parsed data from bounced email 
messages like following.

```ruby
#! /usr/bin/env ruby
require 'sisimai'
v = Sisimai.make('/path/to/mbox')       # or path to Maildir/

# If you want to get bounce records which reason is "delivered", set "delivered"
# option to make() method like the following:
v = Sisimai.make('/path/to/mbox', delivered: true)

if v.is_a? Array
  v.each do |e|
    puts e.class                # Sisimai::Data
    puts e.recipient.class      # Sisimai::Address
    puts e.timestamp.class      # Sisimai::Time

    puts e.addresser.address    # shironeko@example.org # From
    puts e.recipient.address    # kijitora@example.jp   # To
    puts e.recipient.host       # example.jp
    puts e.deliverystatus       # 5.1.1
    puts e.replycode            # 550
    puts e.reason               # userunknown

    h = e.damn                  # Convert to HASH
    j = e.dump('json')          # Convert to JSON string
    puts e.dump('json')         # JSON formatted bounce data
  end
end
```

Convert to JSON
-------------------------------------------------------------------------------
`Sisimai.dump()` method provides feature for getting parsed data as JSON string
from bounced email messages like following.

```ruby
# Get JSON string from parsed mailbox or Maildir/
puts Sisimai.dump('/path/to/mbox')  # or path to Maildir/

# dump() method also accepts "delivered" option like the following code:
puts Sisimai.dump('/path/to/mbox', delivered: true)
```

Read bounce object
-------------------------------------------------------------------------------
The way to read a bounce object retrived from Cloud Email Services as JSON using
their API is the following code. This feature is available at Sisimai 4.20.0 or
later.

```ruby
#! /usr/bin/env ruby
require 'json'
require 'sisimai'

j = JSON.load('{"notificationType"=>"Bounce", "bounce"=>{"...') # JSON String
v = Sisimai.make(j, input: 'json')

if v.is_a? Array
  v.each do |e|
    ...
  end
end
```
As of present, Only Amazon SES and SendGrid are supported.

Callback feature
-------------------------------------------------------------------------------
Beginning with Sisimai 4.19.0, `make()` and `dump()` methods of Sisimai accept
a Lambda (Proc object) in `hook` argument for setting a callback method and
getting the results generated by the method via `Sisimai::Data.catch` method.

```ruby
#! /usr/bin/env ruby
require 'sisimai'
callbackto = lambda do |v|
  r = { 'x-mailer' => '', 'queue-id' => '' }

  if cv = v['message'].match(/^X-Postfix-Queue-ID:\s*(.+)$/)
    r['queue-id'] = cv[1]
  end
  r['x-mailer'] = v['headers']['x-mailer'] || ''
  return r
end

list = ['X-Mailer']
data = Sisimai.make('/path/to/mbox', hook: callbackto, field: list)
json = Sisimai.dump('/path/to/mbox', hook: callbackto, field: list)

puts data[0].catch['x-mailer']      # Apple Mail (2.1283)
```

More information about the callback feature is available at
[Sisimai | How To Parse - Callback](http://libsisimai.org/en/usage/#callback)
Page.

One-Liner
-------------------------------------------------------------------------------

```shell
% ruby -rsisimai -e 'puts Sisimai.dump($*.shift)' /path/to/mbox
```

Output example
-------------------------------------------------------------------------------
![](http://libsisimai.org/static/images/demo/sisimai-dump-02.gif)

```json
[{"recipient": "kijitora@example.jp", "addresser": "shironeko@1jo.example.org", "feedbacktype": "", "action": "failed", "subject": "Nyaaaaan", "smtpcommand": "DATA", "diagnosticcode": "550 Unknown user kijitora@example.jp", "listid": "", "destination": "example.jp", "smtpagent": "Courier", "lhost": "1jo.example.org", "deliverystatus": "5.0.0", "timestamp": 1291954879, "messageid": "201012100421.oBA4LJFU042012@1jo.example.org", "diagnostictype": "SMTP", "timezoneoffset": "+0900", "reason": "filtered", "token": "ce999a4c869e3f5e4d8a77b2e310b23960fb32ab", "alias": "", "senderdomain": "1jo.example.org", "rhost": "mfsmax.example.jp"}, {"diagnostictype": "SMTP", "timezoneoffset": "+0900", "reason": "userunknown", "timestamp": 1381900535, "messageid": "E1C50F1B-1C83-4820-BC36-AC6FBFBE8568@example.org", "token": "9fe754876e9133aae5d20f0fd8dd7f05b4e9d9f0", "alias": "", "senderdomain": "example.org", "rhost": "mx.bouncehammer.jp", "action": "failed", "addresser": "kijitora@example.org", "recipient": "userunknown@bouncehammer.jp", "feedbacktype": "", "smtpcommand": "DATA", "subject": "バウンスメールのテスト(日本語)", "destination": "bouncehammer.jp", "listid": "", "diagnosticcode": "550 5.1.1 <userunknown@bouncehammer.jp>... User Unknown", "deliverystatus": "5.1.1", "lhost": "p0000-ipbfpfx00kyoto.kyoto.example.co.jp", "smtpagent": "Sendmail"}]
```

Sisimai Specification
===============================================================================

Differences between Ruby version and Perl version
-------------------------------------------------------------------------------
The following table show the differences between Ruby version of Sisimai
and Perl version of Sisimai. Information about differences between Sisimai
and bounceHammer are available at
[Sisimai | Differences](http://libsisimai.org/en/diff) page.

| Features                                    | Ruby version   | Perl version  |
|---------------------------------------------|----------------|---------------|
| System requirements                         | Ruby 2.1 - 2.4 | Perl 5.10 -   |
|                                             | JRuby 9.0.4.0- |               |
| Analytical precision ratio(2000 emails)[1]  | 1.00           | 1.00          |
| The speed of parsing email(1000 emails)     | 3.30s          | 2.33s         |
| How to install                              | gem install    | cpanm         |
| Dependencies (Except core modules)          | 1 module       | 2 modules     |
| LOC:Source lines of code                    | 12400 lines    | 9000 lines    |
| The number of tests(spec/,t/,xt/) directory | 108000 tests   | 189600 tests  |
| License                                     | BSD 2-Clause   | BSD 2-Clause  |
| Support Contract provided by Developer      | Coming soon    | Available     |

1. See [./ANALYTICAL-PRECISION](https://github.com/sisimai/rb-Sisimai/blob/master/ANALYTICAL-PRECISION)

Other specification of Sisimai
-------------------------------------------------------------------------------
- [**Parser Engines**](http://libsisimai.org/en/engine)
- [**Bounce Reason List**](http://libsisimai.org/en/reason)
- [**Data Structure of Sisimai::Data**](http://libsisimai.org/en/data)

Contributing
===============================================================================

Bug report
-------------------------------------------------------------------------------
Please use the [issue tracker](https://github.com/sisimai/rb-Sisimai/issues)
to report any bugs.

Emails could not be parsed
-------------------------------------------------------------------------------
Bounce mails which could not be parsed by Sisimai are saved in the repository
[set-of-emails/to-be-debugged-because/sisimai-cannot-parse-yet](https://github.com/sisimai/set-of-emails/tree/master/to-be-debugged-because/sisimai-cannot-parse-yet). 
If you have found any bounce email cannot be parsed using Sisimai, please add
the email into the directory and send Pull-Request to this repository.

Other Information
===============================================================================

Related Sites
-------------------------------------------------------------------------------
* __@libsisimai__ | [Sisimai on Twitter (@libsisimai)](https://twitter.com/libsisimai)
* __libSISIMAI.ORG__ | [Sisimai | The successor to bounceHammer, Library to parse bounce mails](http://libsisimai.org/)

Changes
Revision history for the Ruby version of Sisimai

4.21.0p1 (not tagged yet)
    - Pull-Request #73, Fix codes for initializing a hash element with an empty
      string in Sisimai::MTA::Postfix. Thanks to @MichiakiNakaya.
    - Import pull-request: https://github.com/sisimai/p5-Sisimai/pull/222 from
      p5-Sisimai to improve error message patterns.

4.21.0 Mon, 10 Apr 2017 12:17:22 +0900 (JST)
    - Experimental implementation: new MTA module "Sisimai::MSP::US::GSuite" to 
      parse bounce mails returned from G Suite.
    - Sisimai.make and Sisimai::Message.new methods check "field" argument more
      strictly.
    - Improved Sisimai::SMTP::Status.find(). The method checks whether a found
      value as D.S.N. is IPv4 address or not.
    - Improved code for getting error messages, D.S.N. values, and SMTP reply
      codes in Sisimai::MTA::Postfix.scan() method.
    - Pull-Request #69, Fix some typos. Thanks to @koic.
    - Pull-Request #71, Fix break statement. Thanks to @MichiakiNakaya.
    - Issue #70, All the value of $PATH in Makefile have been fixed to build a
      gem file for JRuby Platform. Thanks to @MichiakiNakaya, @hiroyuki-sato.

4.20.2  Sat, 11 Mar 2017 16:32:48 +0900 (JST)
    - Pull-Request #63 Add some error message patterns from Amazon SES SMTP
      endpoint.
    - Fix regular expression in Sisimai::Message::Email.headers for resolving
      a bug reported from @rdeavila at issue #65.
    - Issue #67, Fix code in Sisimai.make method for reading bounce email data
      from STDIN. Thanks to @marine_dayo.
    - Callback feature improvement: import pull-request from sisimai/p5-Sisimai
      https://github.com/sisimai/p5-Sisimai/pull/210.

4.20.1  Sat, 31 Dec 2016 20:10:22 +0900 (JST)
    - Fix the Java version of Gem file.

4.20.0  Sat, 31 Dec 2016 13:36:22 +0900 (JST)
    - Experimental implementation: New MTA modules for 2 Cloud Email Deliveries,
      Sisimai::CED::US::AmazonSES and Sisimai::CED::US::SendGrid. These modules
      can parse JSON formatted bounce objects and can convert to Sisimai::Data
      object.
    - Format of "smtpagent" in the parsed result has been changed. It includes
      the category name of MTA/MSP modules like "MTA::Sendmail", "MTA::Exim",
      and "MSP::US::SendGrid".
    - The Domain part of dummy email address defined in Sisimai::Address module
      has been changed from "dummy-domain.invalid" to "libsisimai.org.invalid".
    - Sisimai::SMTP.is_softbounce() method has been deleted.
    - Improved code for avoid "Invalid byte sequence in UTF-8 (ArgumentError)"
      error in Sisimai::String.to_plain method reported from M.R.
    - Sisimai works on Ruby 2.4.0.

4.19.0  Tue, 18 Oct 2016 14:19:10 +0900 (JST)
    - Implement a callback feature at Sisimai.make and Sisimai.dump methods.
      More imformation about the feature are available at the following pages:
      - http://libsisimai.org/en/usage#callback
      - http://libsisimai.org/ja/usage#callback
    - Implement Sisimai.match() method: issue #52.
    - Minor bug fix in Sisimai::MSP::US::AmazonSES.scan() method.

4.18.1  Sun, 11 Sep 2016 20:05:20 +0900 (JST)
    - Fix bug in Sisimai::Mail::STDIN.read() method reported at issue #61.
      Thanks to @yaegassy.
    - Fix bug in Sisimai::MIME.qprintd() reported at #60.
    - Improved code related to MIME decoding.
    - Implement Sisimai::String.to_plain() for converting from HTML message to
      plain text before parsing. The method and related codes are imported from
      pull-request #197 at p5-Sisimai.

4.18.0  Mon, 22 Aug 2016 20:40:55 +0900 (JST)
    - Import pull-request https://github.com/sisimai/rb-Sisimai/pull/59 (soft
      bounce improvement) from Perl version of Sisimai.
    - Sisimai::MTA::Exchange has been renamed to Sisimai::MTA::Exchange2003.
    - Implement new MTA module Sisimai::MTA::Exchange2007.

4.17.2  Tue, 26 Jul 2016 21:00:17 +0900 (JST)
    - Issue #174, Implement Sisimai::Rhost::ExchangeOnline for the bounce mail
      from on-premises Exchange 2013 and Office 365.
    - The reason of status code: 4.4.5 is "systemfull".
    - Code improvement at Sisimai::MSP::US::Office365.

4.17.1  Wed, 30 Mar 2016 14:00:22 +0900 (JST)
    - Ported codes from https://github.com/sisimai/p5-Sisimai/pull/180 to fix
      issue https://github.com/sisimai/p5-Sisimai/issues/179 a variable in lib/
      sisimai/mta/exim.rb is not quoted before passing to %r// operator.
    - Fixed serious bug in Sisimai::Mail::Maildir#read method reported at issue
      #55 and #56 by pull-request #57. Thanks to @taku1201.

4.17.0  Wed, 16 Mar 2016 12:22:44 +0900 (JST)
    - Implement new reason "syntaxerror". Sisimai will set "syntaxerror" to the
      raeson when the value of "replycode" begins with "50" such as 502, 503,
      or 504. Imported from https://github.com/sisimai/p5-Sisimai/pull/170.
    - Implement description() method at each class in sisimai/reason/*.rb and
      Sisimai.reason method for getting the list of reasons Sisimai can detect
      and its description: issue #48.
    - Remove unused method Sisimai::Reason->match, issue #49.
    - Some methods of Sisimai::Address allow "postmaster", and "mailer-daemon"
      ( without a domain part ) as an email address.
    - Sisimai::RFC5322.is_mailerdaemon() method returns true when the argument
      includes "postmaster".
    - Merge pull-request #51, new method Sisimai::RFC5322.weedout() and code
      improvements in all the MTA/MSP modules.

4.16.0  Thu, 18 Feb 2016 13:49:01 +0900 (JST)
    - Implement new reason "delivered". Sisimai set "delivered" to the reason
      when the value of Status: field in a bounce message begins with "2". This
      feature is optional and is not enabled by default.
    - Implement new method Sisimai.engine. the method returns the list of MTA
      and MSP module list implemented in Sisimai.
    - Fix serious bug (`gem install` fails on JRuby environment) in Gemfile and

lib/sisimai/data.rb
module Sisimai
  # Sisimai::Data generate parsed data from Sisimai::Message object.
  class Data
    # Imported from p5-Sisimail/lib/Sisimai/Data.pm
    require 'sisimai/address'
    require 'sisimai/rfc5322'
    require 'sisimai/smtp/error'
    require 'sisimai/smtp/reply'
    require 'sisimai/smtp/status'
    require 'sisimai/string'
    require 'sisimai/reason'
    require 'sisimai/rhost'
    require 'sisimai/time'
    require 'sisimai/datetime'

    @@rwaccessors = [
      :catch,           # [?] Results generated by hook method
      :token,           # [String] Message token/MD5 Hex digest value
      :lhost,           # [String] local host name/Local MTA
      :rhost,           # [String] Remote host name/Remote MTA
      :alias,           # [String] Alias of the recipient address
      :listid,          # [String] List-Id header of each ML
      :reason,          # [String] Bounce reason
      :action,          # [String] The value of Action: header
      :subject,         # [String] UTF-8 Subject text
      :timestamp,       # [Sisimai::Time] Date: header in the original message
      :addresser,       # [Sisimai::Address] From address
      :recipient,       # [Sisimai::Address] Recipient address which bounced
      :messageid,       # [String] Message-Id: header
      :replycode,       # [String] SMTP Reply Code
      :smtpagent,       # [String] MTA name
      :softbounce,      # [Integer] 1 = Soft bounce, 0 = Hard bounce, -1 = ?
      :smtpcommand,     # [String] The last SMTP command
      :destination,     # [String] The domain part of the "recipinet"
      :senderdomain,    # [String] The domain part of the "addresser"
      :feedbacktype,    # [String] Feedback Type
      :diagnosticcode,  # [String] Diagnostic-Code: Header
      :diagnostictype,  # [String] The 1st part of Diagnostic-Code: Header
      :deliverystatus,  # [String] Delivery Status(DSN)
      :timezoneoffset,  # [Integer] Time zone offset(seconds)
    ]
    @@rwaccessors.each { |e| attr_accessor e }

    EndOfEmail = Sisimai::String.EOM
    RetryIndex = Sisimai::Reason.retry
    RFC822Head = Sisimai::RFC5322.HEADERFIELDS(:all)
    AddrHeader = {
      addresser: RFC822Head[:addresser],
      recipient: RFC822Head[:recipient],
    }
    ActionList = %r/\A(?:failed|delayed|delivered|relayed|expanded)\z/
    ActionHead = {
      %r/\Afailure\z/ => 'failed',
      %r/\Aexpired\z/ => 'delayed',
    }

    # Constructor of Sisimai::Data
    # @param    [Hash] argvs    Data
    # @return   [Sisimai::Data] Structured email data
    def initialize(argvs)
      thing = {}

      # Create email address object
      x0 = Sisimai::Address.parse([argvs['addresser']])
      y0 = Sisimai::Address.parse([argvs['recipient']])

      return nil unless x0.is_a? Array
      return nil unless y0.is_a? Array

      thing[:addresser] = Sisimai::Address.new(x0.shift)
      return nil unless thing[:addresser].is_a? Sisimai::Address
      return nil if thing[:addresser].void
      thing[:senderdomain] = thing[:addresser].host

      thing[:recipient] = Sisimai::Address.new(y0.shift)
      return nil unless thing[:recipient].is_a? Sisimai::Address
      return nil if thing[:recipient].void
      thing[:destination] = thing[:recipient].host
      thing[:alias] = argvs['alias'] || ''

      @addresser    = thing[:addresser]
      @senderdomain = thing[:senderdomain]
      @recipient    = thing[:recipient]
      @destination  = thing[:destination]
      @alias        = thing[:alias]

      @token = Sisimai::String.token(@addresser.address, @recipient.address, argvs['timestamp'])
      @timestamp = Sisimai::Time.parse(::Time.at(argvs['timestamp']).to_s)
      @timezoneoffset = argvs['timezoneoffset'] || '+0000'
      @lhost          = argvs['lhost']          || ''
      @rhost          = argvs['rhost']          || ''
      @catch          = argvs['catch']          || nil
      @reason         = argvs['reason']         || ''
      @listid         = argvs['listid']         || ''
      @subject        = argvs['subject']        || ''
      @messageid      = argvs['messageid']      || ''
      @smtpagent      = argvs['smtpagent']      || ''
      @diagnosticcode = argvs['diagnosticcode'] || ''
      @diagnostictype = argvs['diagnostictype'] || ''
      @deliverystatus = argvs['deliverystatus'] || ''
      @smtpcommand    = argvs['smtpcommand']    || ''
      @feedbacktype   = argvs['feedbacktype']   || ''
      @action         = argvs['action']         || ''
      @replycode      = argvs['replycode']      || ''
      @replycode      = Sisimai::SMTP::Reply.find(argvs['diagnosticcode']) if @replycode.empty?
      @softbounce     = argvs['softbounce']     || ''
    end

    # Another constructor of Sisimai::Data
    # @param          [Sisimai::Message] data Data Object
    # @param          [Hash] argvs            Parser options
    # @options argvs  [Boolean] delivered     true: Including "delivered" reason
    # @return         [Array, Nil]            List of Sisimai::Data or Nil if the
    #                                         argument is not Sisimai::Message object
    def self.make(data: nil, **argvs)
      return nil unless data
      return nil unless data.is_a? Sisimai::Message

      messageobj = data
      mailheader = data.header
      rfc822data = messageobj.rfc822
      fieldorder = { :recipient => [], :addresser => [] }
      objectlist = []
      rxcommands = %r/\A(?:EHLO|HELO|MAIL|RCPT|DATA|QUIT)\z/
      givenorder = argvs[:order] || {}
      delivered1 = argvs[:delivered] || false

      return nil unless messageobj.ds
      return nil unless messageobj.rfc822
      require 'sisimai/smtp'

      # Decide the order of email headers: user specified or system default.
      if givenorder.is_a?(Hash) && givenorder.keys.size > 0
        # If the order of headers for searching is specified, use the order
        # for detecting an email address.
        fieldorder.each_key do |e|
          # The order should be "Array Reference".
          next unless givenorder[e]
          next unless givenorder[e].is_a? Array
          next unless givenorder[e].size > 0
          fieldorder[e].concat(givenorder[e])
        end
      end

      fieldorder.each_key do |e|
        # If the order is empty, use default order.
        if fieldorder[e].empty?
          # Load default order of each accessor.
          fieldorder[e] = AddrHeader[e]
        end
      end

      messageobj.ds.each do |e|
        # Create parameters for new() constructor.
        p = {
          'catch'          => messageobj.catch  || nil,
          'lhost'          => e['lhost']        || '',
          'rhost'          => e['rhost']        || '',
          'alias'          => e['alias']        || '',
          'action'         => e['action']       || '',
          'reason'         => e['reason']       || '',
          'replycode'      => e['replycode']    || '',
          'smtpagent'      => e['agent']        || '',
          'recipient'      => e['recipient']    || '',
          'softbounce'     => e['softbounce']   || '',
          'smtpcommand'    => e['command']      || '',
          'feedbacktype'   => e['feedbacktype'] || '',
          'diagnosticcode' => e['diagnosis']    || '',
          'diagnostictype' => e['spec']         || '',
          'deliverystatus' => e['status']       || '',
        }
        unless delivered1
          # Skip if the value of "deliverystatus" begins with "2." such as 2.1.5
          next if p['deliverystatus'] =~ /\A2[.]/
        end

        # EMAIL_ADDRESS:
        # Detect email address from message/rfc822 part
        fieldorder[:addresser].each do |f|
          # Check each header in message/rfc822 part
          h = f.downcase
          next unless rfc822data.key?(h)
          next unless rfc822data[h].size > 0
          next unless Sisimai::RFC5322.is_emailaddress(rfc822data[h])
          p['addresser'] = rfc822data[h]
          break
        end

        # Fallback: Get the sender address from the header of the bounced
        # email if the address is not set at loop above.
        p['addresser'] ||= ''
        p['addresser']   = messageobj.header['to'] if p['addresser'].empty?

        next unless p['addresser']
        next unless p['recipient']

        # TIMESTAMP:
        # Convert from a time stamp or a date string to a machine time.
        datestring = nil
        zoneoffset = 0
        datevalues = []
        if e['date'] && e['date'].size > 0
          datevalues << e['date']
        end

        # Date information did not exist in message/delivery-status part,...
        RFC822Head[:date].each do |f|
          # Get the value of Date header or other date related header.
          next unless rfc822data[f.downcase]
          datevalues << rfc822data[f.downcase]
        end

        if datevalues.size < 2
          # Set "date" getting from the value of "Date" in the bounce message
          datevalues << messageobj.header['date']
        end

        datevalues.each do |v|
          # Parse each date value in the array
          datestring = Sisimai::DateTime.parse(v)
          break if datestring
        end

        if datestring
          # Get the value of timezone offset from $datestring
          if cv = datestring.match(/\A(.+)[ ]+([-+]\d{4})\z/)
            # Wed, 26 Feb 2014 06:05:48 -0500
            datestring = cv[1]
            zoneoffset = Sisimai::DateTime.tz2second(cv[2])
            p['timezoneoffset'] = cv[2]
          end
        end

        begin
          # Convert from the date string to an object then calculate time
          # zone offset.
          t = Sisimai::Time.strptime(datestring, '%a, %d %b %Y %T')
          p['timestamp'] = (t.to_time.to_i - zoneoffset) || nil
        rescue
          warn ' ***warning: Failed to strptime ' + datestring.to_s
        end
        next unless p['timestamp']

        # OTHER_TEXT_HEADERS:
        recvheader = mailheader['received'] || []
        if recvheader.size > 0
          # Get localhost and remote host name from Received header.
          %w|lhost rhost|.each { |a| e[a] ||= '' }
          e['lhost'] = Sisimai::RFC5322.received(recvheader[0]).shift if e['lhost'].empty?
          e['rhost'] = Sisimai::RFC5322.received(recvheader[-1]).pop  if e['rhost'].empty?
        end

        # Remove square brackets and curly brackets from the host variable
        %w|rhost lhost|.each do |v|
          p[v] = p[v].delete('[]()')    # Remove square brackets and curly brackets from the host variable
          p[v] = p[v].sub(/\A.+=/, '')  # Remove string before "="
          p[v] = p[v].gsub(/\r\z/, '')  # Remove CR at the end of the value

          # Check space character in each value
          if p[v] =~ / /
            # Get the first element
            p[v] = p[v].split(' ', 2).shift
          end
        end

        # Subject: header of the original message
        p['subject'] = rfc822data['subject'] || ''
        p['subject'] = p['subject'].gsub(/\r\z/, '')

        # The value of "List-Id" header
        p['listid'] = rfc822data['list-id'] || ''
        if p['listid'].size > 0
          # Get the value of List-Id header
          if cv = p['listid'].match(/\A.*([<].+[>]).*\z/)
            # List name <list-id@example.org>
            p['listid'] = cv[1]
          end
          p['listid'] = p['listid'].delete('<>')
          p['listid'] = p['listid'].gsub(/\r\z/, '')
          p['listid'] = '' if p['listid'] =~ / /
        end

        # The value of "Message-Id" header
        p['messageid'] = rfc822data['message-id'] || ''
        if p['messageid'].size > 0
          # Remove angle brackets
          if cv = p['messageid'].match(/\A([^ ]+)[ ].*/)
            p['messageid'] = cv[1]
          end
          p['messageid'] = p['messageid'].delete('<>')
          p['messageid'] = p['messageid'].gsub(/\r\z/, '')
        end

        # CHECK_DELIVERY_STATUS_VALUE:
        # Cleanup the value of "Diagnostic-Code:" header
        p['diagnosticcode'] = p['diagnosticcode'].sub(/[ \t]+#{EndOfEmail}/, '')
        d = Sisimai::SMTP::Status.find(p['diagnosticcode'])
        if d =~ /\A[45][.][1-9][.][1-9]\z/
          # Use the DSN value in Diagnostic-Code:
          p['deliverystatus'] = d
        end

        if p['reason'] == 'mailererror'
          p['diagnostictype'] ||= 'X-UNIX'
        else
          unless p['reason'] =~ /\A(?:feedback|vacation)\z/
            p['diagnostictype'] ||= 'SMTP'
          end
        end

        # Check the value of SMTP command
        p['smtpcommand'] = '' unless p['smtpcommand'] =~ rxcommands

        # Check the value of "action"
        if p['action'].size > 0
          if cv = p['action'].match(/\A(.+?) .+/)
            # Action: expanded (to multi-recipient alias)
            p['action'] = cv[1]
          end

          unless p['action'] =~ ActionList
            # The value of "action" is not in the following values:
            # "failed" / "delayed" / "delivered" / "relayed" / "expanded"
            ActionHead.each_key do |q|
              next unless p['action'] =~ q
              p['action'] = ActionHead[q]
              break
            end
          end
        else
          if p['reason'] == 'expired'
            # Action: delayed
            p['action'] = 'delayed'
          elsif p['deliverystatus'] =~ /\A[45]/
            # Action: failed
            p['action'] = 'failed'
          end
        end

        o = Sisimai::Data.new(p)
        next unless o.recipient

        if o.reason.empty? || RetryIndex.index(o.reason)
          # Decide the reason of email bounce
          r = ''
          if Sisimai::Rhost.match(o.rhost)
            # Remote host dependent error
            r = Sisimai::Rhost.get(o)
          end
          r = Sisimai::Reason.get(o) if r.empty?
          r = 'undefined' if r.empty?
          o.reason = r
        end

        if o.reason =~ /\A(?:delivered|feedback|vacation)\z/
          # The value of reason is "vacation" or "feedback"
          o.softbounce = -1
          o.replycode = '' unless o.reason == 'delivered'
        else
          # Bounce message which reason is "feedback" or "vacation" does
          # not have the value of "deliverystatus".
          softorhard = nil

          if o.softbounce.to_s.empty?
            # The value is not set yet
            textasargv = sprintf('%s %s', p['deliverystatus'], p['diagnosticcode'])
            textasargv = textasargv.gsub(/\A[ ]/, '')
            softorhard = Sisimai::SMTP::Error.soft_or_hard(o.reason, textasargv)

            if softorhard.size > 0
              # Returned value is "soft" or "hard"
              o.softbounce = (softorhard == 'soft') ? 1 : 0
            else
              # Returned value is an empty string
              o.softbounce = (-1)
            end
          end

          if o.deliverystatus.empty?
            # Set pseudo status code
            textasargv = sprintf('%s %s', o.replycode, p['diagnosticcode'])
            textasargv = textasargv.gsub(/\A[ ]/, '')

            getchecked = Sisimai::SMTP::Error.is_permanent(textasargv)
            tmpfailure = getchecked.nil? ? false : (getchecked ? false : true)
            pseudocode = Sisimai::SMTP::Status.code(o.reason, tmpfailure)

            if pseudocode.size > 0
              # Set the value of "deliverystatus" and "softbounce"
              o.deliverystatus = pseudocode

              if o.softbounce < 0
                # set the value of "softbounce" again when the value is -1
                softorhard = Sisimai::SMTP::Error.soft_or_hard(o.reason, pseudocode)

                if softorhard.size > 0
                  # Returned value is "soft" or "hard"
                  o.softbounce = softorhard == 'soft' ? 1 : 0
                else
                  # Returned value is an empty string
                  o.softbounce = -1
                end
              end
            end
          end

          if o.replycode.size > 0
            # Check both of the first digit of "deliverystatus" and "replycode"
            o.replycode = '' unless o.replycode[0, 1] == o.deliverystatus[0, 1]
          end

        end
        objectlist << o

      end
      return objectlist
    end

    # Convert from object to hash reference
    # @return   [Hash] Data in Hash reference
    def damn
      data = {}
      @@rwaccessors.each do |e|
        next if e.to_s =~ /(?:addresser|recipient|timestamp)/
        data[e.to_s] = self.send(e) || ''
      end
      data['addresser'] = self.addresser.address
      data['recipient'] = self.recipient.address
      data['timestamp'] = self.timestamp.to_time.to_i
      return data
    end
    alias :to_hash :damn

    # Data dumper
    # @param    [String] type   Data format: json, yaml
    # @return   [String, Undef] Dumped data or Undef if the value of first
    #                           argument is neither "json" nor "yaml"
    def dump(type = 'json')
      return nil unless %w|json yaml|.index(type)
      referclass = sprintf('Sisimai::Data::%s', type.upcase)

      begin
        require referclass.downcase.gsub('::', '/')
      rescue
        warn '***warning: Failed to load' + referclass
      end

      dumpeddata = Module.const_get(referclass).dump(self)
      return dumpeddata
    end

    # JSON handler
    # @return   [String]            JSON string converted from Sisimai::Data
    def to_json(*)
      return self.dump('json')
    end

  end
end
